pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/decoder/protobufnative/PulsarProtobufNativeRowDecoderFactory.java

```diff
@@ -0,0 +1,163 @@
+        }
+
+        //Protobuf have not yet supported Cyclic Objects.
+        columnMetadata = schema.getFields().stream()
+                .map(field ->
+                        new PulsarColumnMetadata(PulsarColumnMetadata.getColumnName(handleKeyValueType, field.getName()),
+                                parseProtobufPrestoType(field), field.getType().toString(), null, false, false,
+                                handleKeyValueType, new PulsarColumnMetadata.DecoderExtraInfo(field.getName(),
+                                null, null))
+
+                ).collect(toList());
+
+        return columnMetadata;
+    }
+
+    private Type parseProtobufPrestoType(Descriptors.FieldDescriptor field) {
+        //parse by proto JavaType
+        Descriptors.FieldDescriptor.JavaType type = field.getJavaType();
+        Type dataType;
+        switch (type) {
+            case BOOLEAN:
+                dataType = BooleanType.BOOLEAN;
+                break;
+            case BYTE_STRING:
+                dataType = VarbinaryType.VARBINARY;
+                break;
+            case DOUBLE:
+                dataType = DoubleType.DOUBLE;
+                break;
+            case ENUM:
+            case STRING:
+                dataType = createUnboundedVarcharType();
+                break;
+            case FLOAT:
+                dataType = RealType.REAL;
+                break;
+            case INT:
+                dataType = IntegerType.INTEGER;
+                break;
+            case LONG:
+                dataType = BigintType.BIGINT;
+                break;
+            case MESSAGE:
+                Descriptors.Descriptor msg = field.getMessageType();
+                if (field.isMapField()) {
+                    //map
+                    TypeSignature keyType = parseProtobufPrestoType(msg.findFieldByName("key")).getTypeSignature();
+                    TypeSignature valueType = parseProtobufPrestoType(msg.findFieldByName("value")).getTypeSignature();
+                    return typeManager.getParameterizedType(StandardTypes.MAP,
+                            ImmutableList.of(TypeSignatureParameter.typeParameter(keyType),
+                                    TypeSignatureParameter.typeParameter(valueType)));
+                } else {
+                    //row
+                    dataType = RowType.from(msg.getFields().stream()
+                            .map(rowField -> new RowType.Field(Optional.of(rowField.getName()),
+                                    parseProtobufPrestoType(rowField)))
+                            .collect(toImmutableList()));
+                }
+                break;
+            default:
+                throw new RuntimeException("Unknown type: " + type.toString() + " for FieldDescriptor: " + field.getName());
+        }
+        //list
+        if (field.isRepeated() && !field.isMapField()) {
+            dataType = new ArrayType(dataType);
+        }
```

Comment:
I think this logic should move to switch block.

Reply:
Protobuf Array type is not a base type , only mark by Repeated flag , ArrayType Constructor method need the result of switch block .